connect/runtime/src/test/java/org/apache/kafka/connect/util/clusters/EmbeddedConnectCluster.java

```diff
@@ -63,24 +71,58 @@
     private static final Properties DEFAULT_BROKER_CONFIG = new Properties();
     private static final String REST_HOST_NAME = "localhost";
 
-    private final Connect[] connectCluster;
+    private final List<Connect> connectCluster;
     private final EmbeddedKafkaCluster kafkaCluster;
     private final Map<String, String> workerProps;
     private final String connectClusterName;
     private final int numBrokers;
+    private final int numInitialWorkers;
+    private final boolean failOnUngracefulShutdown;
 
-    private EmbeddedConnectCluster(String name, Map<String, String> workerProps, int numWorkers, int numBrokers, Properties brokerProps) {
+    private EmbeddedConnectCluster(String name, Map<String, String> workerProps, int numWorkers,
+                                   int numBrokers, Properties brokerProps,
+                                   boolean failOnUngracefulShutdown) {
         this.workerProps = workerProps;
         this.connectClusterName = name;
         this.numBrokers = numBrokers;
         this.kafkaCluster = new EmbeddedKafkaCluster(numBrokers, brokerProps);
-        this.connectCluster = new Connect[numWorkers];
+        this.connectCluster = new ArrayList<>();
+        this.numInitialWorkers = numWorkers;
+        this.failOnUngracefulShutdown = failOnUngracefulShutdown;
     }
 
+    /**
+     * A more graceful way to handle abnormal exit of services in integration tests.
+     */
+    public Exit.Procedure exitProcedure = (code, message) -> {
+        if (code != 0) {
+            String exitMessage = "Abrupt service exit with: " + code + " and " + message;
+            log.warn(exitMessage);
+            throw new UngracefulShutdownException(exitMessage);
+        }
+        Exit.exit(0, message);
+    };
+
+    /**
+     * A more graceful way to handle abnormal halt of services in integration tests.
+     */
+    public Exit.Procedure haltProcedure = (code, message) -> {
+        if (code != 0) {
+            String haltMessage = "Abrupt service halt with: " + code + " and " + message;
+            log.warn(haltMessage);
+            throw new UngracefulShutdownException(haltMessage);
+        }
+        Exit.halt(0, message);
+    };
+
     /**
      * Start the connect cluster and the embedded Kafka and Zookeeper cluster.
      */
     public void start() throws IOException {
+        if (!failOnUngracefulShutdown) {
```

Comment:
The two procedures will throw an `UngracefulShutdownException` if the code is non-zero, and according to this block these procedures will be used if `failOnUngracefulShutdown == false`, which seems contradictory to the expectations of the `failOnUngracefulShutdown` name. Am I missing something?

Reply:
The code is a bit more involved :) 

Fail here refers to calls to `Runtime.getRuntime().halt(statusCode);` and `System.exit(statusCode);` in `halt` and `exit` methods respectively. If we don't override those, we get error codes that the test interprets as failures, even if our intention is not to interpret unclean shutdown as failure. 

However, my intention is not to ignore an abnormal shutdown completely, but rather change what's thrown. So, yes, we now throw a runtime exception (`UngracefulShutdownException`) which can be handled by the embedded cluster. I could dig in a bit more to see why the test doesn't fail when the workers is not cleanly shutdown (which I observed in the logs). But I could add a specialized `catch` for `UngracefulShutdownException`, which I forgot I guess.

Reply:
After digging a bit deeper on this case, turns out throwing an exception instead of calling exit worked because the exception is failing the herder executor's thread and we don't use the future. Thus any exception there is swallowed (as opposed to `Exit.exit(non-zero-status)`)

The suggestion in the PRs current form is to handle `UngracefulShutdownException` just by printing a warning and not by failing the test.